### PR TITLE
Template and macros typing support (4.0)

### PR DESCRIPTION
**Pull request by bazsi**
This branch is a continuation of #3885 and adds support for types when using template expressions (and macros). With just #3885 we only get proper types, if the value pairs based functionality references name-value pairs directly (as opposed to templates and macros).

So, with #3885 only and ${number} being a typed name-value pair, this is typed: ```$(format-json number)```
This results in string values: ```$(format-json foo=$number)``` (as we are using $number through a template string).

This patch would change the second form to emit integers instead of strings.

The same also works for macros.

Some of the macros are yielding typed values, I've concentrated all these changes in a single commit and wrote a detailed explanation in the commit message.

Compatibility:
  * in 3.x mode, everything remains a string, except with an explicit --no-cast option in value-pairs based functionality.

Call sites using this which might need compatibility/upgrade warnings (those using types today):
- [x] filter expressions: right now comparisons use the operator to switch between numeric/string comparisons, there are patches to take types into account, but compatibility needs to be handled there. template() options don't need to suppor types, all other filtering is string based.
- [x] mongodb: using value-pairs which is already covered. The collection() parameter uses templates  but that's a string.
- [x] basicfuncs
  - [x] $(grep): string based, no need for explicit support
  - [x] $(if): condition covered by filter expressions, result expansion propagates types already
  - [x] $(context-lookup): returns list of strings
  - [x] $(context-values): returns list of strings
  - [x] numeric operators (+, -, /, * etc): ignores input type, attempts numeric conversion, output is typed. TODO: input parsing should be migrated to generic-number instead of a local, similar implementation here
  - [x] $(iterate): template input (and a literal string), when the template is formatted, its output type is handled and is the result of $(iterate)
  - [x] $(map): input is a list, elements are used as strings (and we don't support typed list elements atm), output is a list, again of strings. No type support possible.
  - [x] $(template): returns the type of the invoked template as it should.
- [x] dbparser: db-parser() emits type information but does not make use of it.
- [x] json: covered via value-pairs, JSON does not on its own use LogTemplate instances, only through value-pairs which already has compatibility functions with 3.x
- [x] riemann: riemann sets various fields using LogTemplates, most of which are strings, with the only exception is time. The time field is explicitly converted using g_ascii_strtoll(), so it ignores the type information return by the template. Compatibility wise this is OK. TODO: error handling in the conversion is completely missing.
- [x] python: python is not typed in its python-logtemplate wrapper. TODO: implement Python typing. Currently it is compatible with 3.x even if typing is enabled.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by MrAnno** [comment]
LGTM. A test is failing due a recently merged template function, which needs to be followed up: `tf_value_pairs_call`.

I'm going to test the PR manually tomorrow to try the new typed macros and templates, otherwise: 👍

**Comment by bazsi** [comment]
I've fixed up the failing unit test caused by a parallel merge to master, so this should be ready to go in if the tests are green (they are for me locally)

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Review by MrAnno — APPROVED** [review]

File: modules/basicfuncs/cond-funcs.c [hunk 1]
```diff
@@ -112,13 +112,15 @@ tf_grep_prepare(LogTemplateFunction *self, gpointer s, LogTemplate *parent, gint
...
-tf_grep_call(LogTemplateFunction *self, gpointer s, const LogTemplateInvokeArgs *args, GString *result)
+tf_grep_call(LogTemplateFunction *self, gpointer s, const LogTemplateInvokeArgs *args, GString *result,
+             LogMessageValueType *type)
 {
   gint i, msg_ndx;
   gboolean first = TRUE;
   TFCondState *state = (TFCondState *) s;
   gint count = 0;
 
+  *type = LM_VT_STRING;
```
[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/basicfuncs/cond-funcs.c:123]: @Kokan Do you consider this note as something that's worth changing here in this PR, or could we do it separately?

[the remark this answers is not in the document]
Comment (Kokan) [reply modules/basicfuncs/cond-funcs.c:123]: We should resolve it here. I would consider merging the PR without resolving a comment if that blocks anything critical (like build, or huge bug)

[the remark this answers is not in the document]
Comment (MrAnno) [reply modules/basicfuncs/cond-funcs.c:123]: Yeah I know. I was just curious about whether it is optional or not. Thanks.

[the remark this answers is not in the document]
Comment (bazsi) [reply modules/basicfuncs/cond-funcs.c:123]: $(grep) predates list support and the format it produces is not list compatible (does not escape values properly). When the time has come to fix it, I opted to create a new names, e.g. $(context-values) and $(context-lookup) and left $(grep) alone. I have doubts that anyone would be using $(grep), but I haven't touched it back then and didn't touch it now. That's why it is left as LM_VT_STRING.

I could deprecate the function, or simply change it to be an alias for $(context-lookup)

**Review by Kokan — APPROVED** [review]

**Pull request merged**